content/oci/oci.go

```diff
@@ -0,0 +1,252 @@
+		return err
+	}
+
+	return s.graph.Index(ctx, s.storage, expected)
+}
+
+// Exists returns true if the described content exists.
+func (s *Store) Exists(ctx context.Context, target ocispec.Descriptor) (bool, error) {
+	return s.storage.Exists(ctx, target)
+}
+
+// Tag tags a descriptor with a reference string.
+// A reference should be either a valid tag (e.g. "latest"),
+// or a digest matching the descriptor (e.g. "@sha256:abc123").
+// Reference: https://github.com/opencontainers/image-spec/blob/main/image-layout.md#indexjson-file
+func (s *Store) Tag(ctx context.Context, desc ocispec.Descriptor, reference string) error {
+	if err := validateReference(reference); err != nil {
+		return err
+	}
+
+	exists, err := s.storage.Exists(ctx, desc)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return fmt.Errorf("%s: %s: %w", desc.Digest, desc.MediaType, errdef.ErrNotFound)
+	}
+
+	if desc.Annotations == nil {
+		desc.Annotations = map[string]string{}
+	}
+	desc.Annotations[ocispec.AnnotationRefName] = reference
+
+	if err := s.resolver.Tag(ctx, desc, reference); err != nil {
+		return err
+	}
+
+	if s.AutoSaveIndex {
+		return s.SaveIndex()
+	}
+	return nil
+}
+
+// Resolve resolves a reference to a descriptor.
+func (s *Store) Resolve(ctx context.Context, reference string) (ocispec.Descriptor, error) {
+	if reference == "" {
+		return ocispec.Descriptor{}, errdef.ErrMissingReference
+	}
+
+	return s.resolver.Resolve(ctx, reference)
+}
+
+// UpEdges returns the nodes directly pointing to the current node.
+// UpEdges returns nil without error if the node does not exists in the store.
+func (s *Store) UpEdges(ctx context.Context, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	return s.graph.UpEdges(ctx, node)
+}
+
+// ensureOCILayoutFile ensures the `oci-layout` file.
+func (s *Store) ensureOCILayoutFile() error {
+	layoutFilePath := filepath.Join(s.root, ocispec.ImageLayoutFile)
+	layoutFile, err := os.Open(layoutFilePath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("failed to open OCI layout file: %v", err)
+		}
+
+		layout := ocispec.ImageLayout{
+			Version: ocispec.ImageLayoutVersion,
+		}
+		layoutJSON, err := json.Marshal(layout)
+		if err != nil {
+			return fmt.Errorf("failed to marshal OCI layout file: %v", err)
+		}
+
+		return ioutil.WriteFile(layoutFilePath, layoutJSON, 0666)
+	}
+	defer layoutFile.Close()
+
+	var layout *ocispec.ImageLayout
+	err = json.NewDecoder(layoutFile).Decode(&layout)
+	if err != nil {
+		return fmt.Errorf("failed to decode OCI layout file: %v", err)
+	}
+	if layout.Version != ocispec.ImageLayoutVersion {
+		return errdef.ErrUnsupportedVersion
+	}
+
+	return nil
+}
+
+// loadIndex reads the index.json from the file system.
+func (s *Store) loadIndex(ctx context.Context) error {
+	indexFile, err := os.Open(s.indexPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("failed to open index file: %v", err)
```

Comment:
nit: it will be better to use `%w` instead of `%v` for wrapping errors. Same comment will be applied to all similar occourances.

Reply:
Please see [fmt.Errorf()](https://pkg.go.dev/fmt#Errorf) for the documentation of `%w`.

Reply:
> What if the user invokes s.Tag(ctx, desc, "localhost:5000/myrepo:mytag") since it is a common scenario in Copy(). Will you parse it and only store mytag in org.opencontainers.image.ref.name? or consider localhost:5000/myrepo:mytag as an invalid reference? 

We need to store the original user-provided reference in the `org.opencontainers.image.ref.name` as we use it to construct ref map when loading index. As per the [OCI image spec](https://github.com/opencontainers/image-spec/blob/main/annotations.md), a reference like "localhost:5000/myrepo:mytag" is valid in grammar.
For the common `Copy` scenarios, isn't it like
```go
// do something
ref := "mytag"
err := oci.Push(ctx, desc, content)
err := oci.Tag(ctx, desc, ref)

repo := "localhost:5000/myrepo"
repository, err := remote.NewRepository(repo)
// do something
root, err := oras.Copy(ctx, oci, ref, repository, ref)
```
? The user would call `oci.Tag(ctx, desc, "mytag") ` instead of `oci.Tag(ctx, desc, "localhost:5000/myrepo:mytag") `

Reply:
How about this scenario?
```golang
ref := "localhost:5000/hello:v1"
src, _ := remote.NewRepository(ref)
dst, _ := oci.New("hello")
oras.Copy(context.Background(), src, ref, dst, "")
```

Reply:
I think it's ok. In this case we will have `"org.opencontainers.image.ref.name"`: `"localhost:5000/hello:v1"`.
The [OCI image spec](https://github.com/opencontainers/image-spec/blob/main/image-layout.md#indexjson-file) mentions that
> No semantic restriction is given for the "org.opencontainers.image.ref.name" annotation of descriptors.
So it's make sense to store FQDN in `org.opencontainers.image.ref.name`.

In terms of API usage, users can still do 
```go
desc, _ := dst.Resolve(ctx, "localhost:5000/hello:v1")
```